src/utils/formContext.ts

```diff
@@ -14,101 +14,113 @@ import { validateElements } from './validation';
+    setFormInternalState(formUuid, {
+      fields: {}
     });
-  },
-  // @deprecated
-  // TODO: remove when support setFieldOptions is dropped
-  setFieldOptions: (newOptions: FieldOptions) => {
-    console.warn(
-      'setFieldOptions is deprecated.  Please use the fields object instead and set the options directly on individual fields.'
-    );
-    const { steps, updateFieldOptions, currentStep } = internalState[formUuid];
-    return updateFieldOptions(steps, currentStep)(newOptions);
-  },
-  setProgress: (val: any) => {
-    return internalState[formUuid].setUserProgress(val);
-  },
-  goToStep: (stepKey: any) => {
-    const { currentStep, history, steps } = internalState[formUuid];
-    changeStep(stepKey, currentStep.key, steps, history);
-  },
-  isLastStep: () => {
-    const step = internalState[formUuid].currentStep;
-    return step.next_conditions.length === 0;
-  },
-  getStepProperties: () => {
-    const state = internalState[formUuid];
-    const step = state?.currentStep ?? {};
-    const rootStyles = step
-      ? step.subgrids.find((grid: any) => grid.position.length === 0).styles
-      : {};
-    return {
-      totalSteps: Object.keys(state.steps).length,
-      stepName: step.key ?? '',
-      previousStepName: state.previousStepName,
-      backgroundColor: rootStyles?.background_color ?? 'FFFFFF'
-    };
-  },
-  // @deprecated
-  // TODO: remove when support setFieldOptions is dropped
-  getFormFields: () => {
-    console.warn(
-      'getFormFields is deprecated.  Please use the fields object instead.'
-    );
-    return formatAllFormFields(internalState[formUuid].steps, true);
-  },
-  // @deprecated
-  // TODO: remove when support setFieldOptions is dropped
-  setFieldErrors: (
-    errors: Record<string, string | { index: number; message: string }>
-  ) => {
-    console.warn(
-      'setFieldErrors is deprecated.  Please use the fields object instead and set the error directly on a field.'
-    );
-    return internalState[formUuid].setFieldErrors(errors);
-  },
-  validateStep: (showErrors = true) => {
-    const {
-      currentStep,
-      visiblePositions,
-      formRef,
-      formSettings,
-      getErrorCallback,
-      setInlineErrors
-    } = internalState[formUuid];
-
-    const { errors } = validateElements({
-      step: currentStep,
-      visiblePositions,
-      triggerErrors: showErrors,
-      errorType: formSettings.errorType,
-      formRef,
-      errorCallback: getErrorCallback(),
-      setInlineErrors
-    });
-    return errors;
   }
-});
+
+  return {
+    userId: initInfo().userId,
+    // deprecated
+    setFieldValues: (userVals: FieldValues): void => {
+      console.warn(
+        'setFieldValues is deprecated.  Please use the fields object and set the value directly in individual fields instead.'
+      );
+      return setFieldValues(userVals);
+    },
+    fields: internalState[formUuid]?.fields ?? {}, // important - should never be undefined
```

Comment:
The nullish coalescing operator is ok here and required because the new setFormInternalState needs to assign an object's properties.  If I get rid of it here, then I would need to change setFormInternalState.  Also, I have tested pretty thoroughly and know it is solid.

Reply:
The internalState.fields object has to be passed directly on each callback context so that when it is eventually loaded it is reflected in context objects that were already returned to the caller/callback.  It has the be the same object.  And yes, callers could "mess" with fields by modifying the object in any number of ways.  The only thing I can think of to do would be whenever we create a new form context (getFormContext is called), we recreate the field wrappers and refresh the internalState.fields object with them and only them.  We would need to pass servar and hidden field keys to it.  This does not really prevent changes though so I don't think we should bother with it.

Reply:
Ditched the nullish

Reply:
Object sealing as well